### DISCUSSION TITLE
Addressing file size limits over federation (different home server limits)

### DISCUSSION THREAD

**Discussion by Palid** [element-hq/element-meta#68] [Ideas]
Currently every single homeserver can have their own size limits for uploaded files. For instance, matrix.org has a size limit of 100MBs, in which case, if you were to upload a file over 100MBs via Element, you'd see a modal saying that the file's too big.

There's a few related problems with current file uploads within federation:

1: How do we handle a case where we have users connecting from different homeservers that has each their own file size limits and someone uploads a file that's over the limit for all of the homeservers, except one?
2: Do we block a user in the same situation from uploading a file that's over their own homeserver file limit, even though a different homeserver in the same room has different limits?
3: Should we set size limits down to the lowest file size limit in the room, or do we allow users to upload files within their limits?
4: If we allow users to upload files within their specific homeserver, how do we handle files that go over our own homeserver limit, but does not go over limit of a different homeserver (e.g. some other user from a different homeserver uploads 200MBs file, where my homeserver supports only up to 100MBs)?

**Comment by Sleuth56** [comment]
The solution I would like to see is. Files sent that are larger than your homeserver's limit will show a message to the user saying something like "This file is over your homeserver's file limit and thus is not stored on your homeserver. If you still want to download this file [click here](otherhomeserver.com/file)." If your homeserver's limit is high enough to accept the file would replicate that file over federation like normal. The download link shown to users would be to the original uploader's homeserver by default.

Possible issues:
1: If the original uploader leaves the room and thus his/her homeserver isn't participating in the room the file could become unavailable.
2: If the original uploader leaves the room but other homeservers have a copy of the file how would those servers announce to the room that they have a copy of the file and if multiple homeservers announce possession of the file how would one be selected as the new default download server.

Possible solutions:
2: When the original uploader leaves a room and his/her server isn't participating in the room each homeserver that has a copy of the file would announce that it has the file. Clients would then have a list of homeservers and would pick one at random to download the file from. 

Room settings:
Room admins could make every file uploaded behave like (See Possible solutions #2). That way room admins could prevent smaller HS's from getting DOSed by users trying to download large files. Spreading out the load amongst the other participating servers instead.

Server settings:
Server's could have a setting to not announce that they have copies of files in rooms.
